New device selection modal

virtuacoplenny opened the pull request:
~~Styling is still in flight a bit and the final test sound for audio output has not been created.~~

virtuacoplenny commented on package.json:
```diff
@@ -55,6 +56,7 @@
     "retry": "0.6.1",
     "strophe": "1.2.4",
     "strophejs-plugins": "0.0.7",
+    "styled-components": "1.3.0",
```
This is a peer-dependency for @atlaskit/util-shared-styles, which is a dependency of @atlaskit/single-select. Without it explicitly added in package.json, the "make" command would fail to properly create the app bundle.

saghul replied:
Hum, probably a good idea to bring up the issue in `@atlaskit/util-shared-styles` ?

virtuacoplenny replied:
For now I've pulled the styled-components changes out into a separate commit and moved it before my changes. I've messaged the maintainer of single-select, asking what the expectation is here. Instead of making a pull request now with the styled-components commit like I said I would, I'm going to hold off until getting an answer.

saghul commented:
Can you drop the styled components commit? It already landed :-)

virtuacoplenny commented:
The commit has been removed and I rebased on top of latest master.

saghul commented:
Thanks Lenny! I'm afraid I ran out of time today, The Real World is awaiting, I'll take a look next week if nobody beats me to it.

saghul commented:
Dome observations before diving into the code (tested on Chrome and Firefox):

- The button label is "video settings", but I'd say "device settings" is more appropriate
- Ditto for the dialog title
- When there is a single entry in the video input combo it says "no matches found" when clicking
- The button labels sound weird to me. "no thanks" should probably be "Cancel" and "submit" -> "Ok" or "Set"

@yanas PTAL at the above.

saghul commented on css/modals/device-selection/_device-selection.scss:
```diff
@@ -0,0 +1,68 @@
+        height: 148px;
+        margin-bottom: 5px;
+
+        .video-input-preview {
```
are we mirroring the video preview? we probably should, since that's how the local view looks like.

virtuacoplenny replied:
Yeah, the mirroring is done with the existing flipVideoX class in use for LocalVideo.js.

saghul commented on react/features/base/devices/reducer.js:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Listen for actions which changes the state of known and used devices.
+ *
+ * @param {Object[]} state - Current state.
```
what are those []? array of object?

virtuacoplenny replied:
I think I meant an Object of arrays but realized I should be more clear and also change the description to match what's used elsewhere for reducer descriptions.

saghul commented on react/features/base/devices/actions.js:
```diff
@@ -0,0 +1,79 @@
+ * }}
+ */
+export function setVideoInputDevice(deviceId) {
+    APP.UI.emitEvent(UIEvents.VIDEO_DEVICE_CHANGED, deviceId);
```
why not use a middleware for this?

virtuacoplenny replied:
It didn't even cross my mind. I've created middleware for this now, which would likely be used anyway when the actual device switching is brought into the web react app.

saghul commented on react/features/device-selection/components/AudioOutputPreview.js:
```diff
@@ -0,0 +1,146 @@
+     * @returns {void}
+     * @private
+     */
+    _setAudioEl(el) {
```
please use full names: setAudioElement

saghul commented on react/features/device-selection/components/AudioOutputPreview.js:
```diff
@@ -0,0 +1,146 @@
+     * @returns {void}
+     * @private
+     */
+    _setSink() {
```
setAudioSink

saghul commented on react/features/device-selection/components/AudioOutputPreview.js:
```diff
@@ -0,0 +1,146 @@
+        if (this._audioEl) {
+            this._audioEl.pause();
+            this._audioEl.removeAttribute('src');
+            this._audioEl.load();
```
is this really necessary?

virtuacoplenny replied:
Thinking about it more, maybe not. I was initially inspired to put in this cleanup because of previous audio switching I did for another project but the use case was different. I've now removed this logic.

saghul commented on react/features/device-selection/components/DeviceSelection.js:
```diff
@@ -0,0 +1,587 @@
+        /**
+         * If true, the audio meter will not display. Necessary for
+         * browsers or configurations that do not support local stats
+         * to prevent a non-responsive mic preview from display.
```
s/display/displaying/

saghul commented on react/features/device-selection/components/DeviceSelector.js:
```diff
@@ -0,0 +1,185 @@
+         * The Select component will adjust the top margin if this instance
+         * is the first of many.
+         */
+        isFirstChild: React.PropTypes.bool,
```
can we avoid having this? looks a bit weird IMHO

virtuacoplenny replied:
By default, selectors will have 20px top margin if isFirstChild is falsy. I've made a change just now so this is hardcoded to true to prevent Selector from adding 20 top margin and using jitsi-meet css to adjust the margins between Selector instances.

saghul commented on react/features/device-selection/components/VideoInputPreview.js:
```diff
@@ -0,0 +1,236 @@
+     * @returns {void}
+     * @private
+     */
+    _attachTrack(newTrack, oldTrack) {
```
I think this function should just take the new track and the caller should be responsible from detaching the previous one.

saghul commented:
Did a first review round. I'm not sure how much I'm going to have this upcoming week, so we will need an extra pair of eyes here.

virtuacoplenny commented:
I changed the audio file to ring.wav and I changed the default "no matches found" string for the case where there are no other devices to select from. I haven't touched the copy; I'll wait for confirmation on the desired copy. I also went ahead and rebased again.

virtuacoplenny commented on react/features/base/devices/reducer.js:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Converts an array of media devices into an object organized by device kind.
+ *
+ * @param {Array} devices - Available media devices.
```
I ended up going back and updating the docs so that devices are characterized as MediaDeviceInfo instances, which is the type returned by WebRTC's enumerateDevices call.

yanas commented:
The PR works really well for me and looks very good. Here are some general remarks, that can be taken care of in a separate PR:
- It looks weird to me that the selected value isn't part of the listed dropdown values. I would expect to see it selected in the dropdown list. Is this some kind of limitation with the AtlasKit component or is it a design choice? The bonus if we do it this way will be that you don't need the "No other choices available" item anymore (the small font of this one looks a little bit off)
- The font of the buttons looks smaller than the dropdown font. Is this a default or a design choice?
FireFox:
- Is there a path on FireFox that would allow changing the microphone even if a reload is needed? If we leave the audio input greyed out then we should add a warning label below the dropdown that would indicate to the user where they can change it in FF.

yanas commented on lang/main.json:
```diff
@@ -422,5 +422,15 @@
+        "noThanks": "No thanks",
+        "selectADevice": "Select a device",
+        "testAudio": "Test sound",
+        "videoInput": "Video input"
```
We need to remove the previously used strings. I believe they are under  "settings":
"selectCamera": "Camera",
"selectMic": "Microphone",
"selectAudioOutput": "Audio output",
and anything else that was used in the previous UI and is no more used here.

yanas commented on react/features/device-selection/components/VideoInputPreview.js:
```diff
@@ -0,0 +1,203 @@
+     */
+    _showMuteOverlay(shouldShow) {
+        if (shouldShow) {
+            this._rootElement.classList.add(VIDE_MUTE_CLASS);
```
Type error VIDE_MUTE_CLASS instead of VIDEO_MUTE_CLASS

yanas commented on react/features/device-selection/components/DeviceSelection.js:
```diff
@@ -0,0 +1,586 @@
+ *
+ * @extends Component
+ */
+class DeviceSelection extends Component {
```
Let's name this one DeviceSelectionDialog, so that it's clear that it's the dialog component and is easy to distinguish it from DeviceSelector component

saghul commented on conference.js:
```diff
@@ -1030,6 +1033,15 @@ export default {
+     *
+     * @returns {JitsiLocalTrack}
+     */
+    getLocalVideoTrack() {
```
do we really need these one liners which are only used once?

virtuacoplenny replied:
I could access it directly but then I'd be accessing conference._room, which is supposed to be private and also has a note saying not to use it. The room variable within conference is private by scope.

saghul commented on css/modals/device-selection/_device-selection.scss:
```diff
@@ -0,0 +1,74 @@
+            background: #0052cc;
+            border-radius: 5px;
+            height: 100%;
+            -webkit-transition: width .1s ease-in-out;
```
Non actionable note: I wonder if there is an equivalent to gulp-autoprefixer which matches our workflow so we didn't have to do this.

saghul commented on lang/main.json:
```diff
@@ -422,5 +422,12 @@
+    "deviceSelection": {
+        "deviceSettings": "Device settings",
+        "noOtherDevices": "No other devices available",
+        "noThanks": "No thanks",
```
@yanas we should pick a word for "cancel" and use it everywhere

yanas replied:
@saghul+++++ I haven't noticed that and it's actually my fault that I haven't named the buttons correctly on the mockups. It should be "Cancel". Lenny can you use cancel instead? Thanks!

saghul commented on react/features/base/devices/actionTypes.js:
```diff
@@ -0,0 +1,45 @@
+import { Symbol } from '../react';
+
+/**
+ * The type of the Redux action which signals that the currently used audio
```
extra "the"

saghul commented on react/features/base/devices/middleware.js:
```diff
@@ -0,0 +1,34 @@
+ * @returns {Function}
+ */
+// eslint-disable-next-line no-unused-vars
+MiddlewareRegistry.register(store => next => action => {
```
❤️

saghul commented on react/features/base/devices/reducer.js:
```diff
@@ -0,0 +1,63 @@
+
+import { ReducerRegistry } from '../redux';
+
+const defaultState = {
```
capitalize global constants

saghul commented on react/features/base/devices/reducer.js:
```diff
@@ -0,0 +1,63 @@
+        }
+
+        // Changing of current audio and video device id is currently handled
+        // outside of react/redux. Fall through to default logic for now.
```
can you put a TODO / FIXME? editors tend to highlight those.

saghul commented on react/features/base/devices/reducer.js:
```diff
@@ -0,0 +1,63 @@
+ * Converts an array of media devices into an object organized by device kind.
+ *
+ * @param {Array<MediaDeviceInfo>} devices - Available media devices.
+ * @private
```
can you check what's the convention we settled on? does the visibility go before or after returns?

lyubomir replied:
If I'm merging a PR, the visibility goes according to alphabetical order as it's above ;)

saghul replied:
Thanks for clarifying @lyubomir!

saghul commented on react/features/base/lib-jitsi-meet/functions.js:
```diff
@@ -60,3 +60,45 @@ export function loadConfig(host: string, path: string = '/config.js') {
+ * @param {string} deviceId - The id of the target video source.
+ * @returns {Promise<JitsiLocalTrack>}
+ */
+export function createLocalVideoTrackPreview(deviceId) {
```
this function is onlt used in one place, do we really need to export 2 of them? can't we export some `createLocalPreviewTrack` instead? Which then brings the question: there is nothing related to a preview here. It's just a local track. I think this could be simplidied.

saghul commented on react/features/base/lib-jitsi-meet/functions.js:
```diff
@@ -60,3 +60,45 @@ export function loadConfig(host: string, path: string = '/config.js') {
+ * @returns {Promise<JitsiLocalTrack>}
+ * @private
+ */
+function createLocalTrack(type, deviceId) {
```
why take an array if we are only using it with a single item?

virtuacoplenny replied:
I'll make a change to expose this function and have it expect a string for type.

virtuacoplenny commented:
I've rebased again onto latest master.
